Collect and display execution metadata for ES|QL cross cluster searches

quux00 opened the pull request:
Early code draft for initial feedback.

Addresses: https://github.com/elastic/elasticsearch/issues/112402

quux00 committed:
Support _cluster/details execution info for ES|QL cross cluster search

Created EsqlExecutionInfo and its inner classes to mirror SearchResponse.Clusters and SearchResponse.Cluster
from the _search side.  EsqlExecutionInfo has been woven into some of the EsqlSession and field-caps API so far.
The enrich pathways are placeholders only for now.

The field-caps response is where the EsqlExecutionInfo gets populated with cluster info (and only
for cross-cluster query scenarios).

quux00 committed:
Added new TransportVersion to support new fields in ComputeResponse, tracking per-cluster:
took time, total shards, successful shards, skipped shards, failed shards

Fixed bug in IndexResolver where the local cluster was not getting added to EsqlExecutionInfo
from the FieldCapsResponses.

Created new ActionListener<ComputeResponse> acquireCompute(String clusterAlias) method in
ComputeListener, but new functionality not yet added to handle results from remote cluster searches.

quux00 committed:
Removed EsqlExecutionInfo from ComputeListener constructor.

Instead it is passed into ActionListener<ComputeResponse> acquireCompute(String clusterAlias, EsqlExecutionInfo executionInfo).
The local cluster metadata is now gathered from startComputeOnDataNodes when clusterAlias="". EsqlExecutionInfo is now passed
into that method (null for remote clusters, non-null for local).

Shard info is set in the EsqlExecutionInfo for all clusters now, but it is all zeros, since the SearchShardsAPI results are
not being added yet - do that in next commit?

quux00 committed:
Now have the remote cluster reporting back it's correct values.
The key finding is that the ClusterListener ctor refs code is what executes
remotely (to send back the ComputeResponse), so you need to pass in EsqlExecutionInfo
into the constructor.

The problem is that the local compute is not getting the shardInfo in the FINAL
ComputeResponse so I think it's the same problem - need to pass EsqlExecutionInfo
into the ClusterListener constructor, perhaps everywhere?

quux00 committed:
I tried putting in exclusions to only gather exec info data for CCS

but I think this is flawed - as we need to gather shard info from other nodes
on the same cluster, both local and remote.

So next attempt: going to use EsqlExecution everywhere for all pathways and
then see if I can remove it from some in the future.

This probably means removing the ComputeListener ctor and acquireCompute methods
that do NOT take clusterAlias and EsqlExecutionInfo

quux00 committed:
Created acquireCCSCompute in ComputeListener and it seems to work now

Cleaned up handling in acquireCCSCompute and acquireCompute for the various
scenarios and now the shard counts are getting properly recorded for remote
clusters on the coordinating cluster.

Next changes: remove EsqlExecutionInfo from acquireCompute and determine
whether we actually need EsqlExecutionInfo in the DataNodeRequestHandler path

quux00 committed:
Removed EsqlExecutionInfo from the DataNodeRequestHandler path

quux00 committed:
Docs update and removed EsqlExecutionInfo from ComputeListener.acquireCompute

quux00 committed:
First pass at adding overallTookTime into Configuration and EsqlExecutionInfo

quux00 committed:
Added tookTime handling for individual clusters.

ActionListener<ComputeResponse> acquireCompute(String clusterAlias)
changed to
ActionListener<ComputeResponse> acquireComputeForDataNodes(String clusterAlias, Configuration configuration)
and now can record per cluster total took time - which is calculated on the remote cluster
not the coordinating cluster

quux00 committed:
EsqlExecutionInfo implements Writeable - not yet tested though

EsqlExecutionInfo added to EsqlQueryResponse, but not to the XContent yet

quux00 committed:
Added overall took time to EsqlQueryResponse

quux00 committed:
Added _cluster/details to Rest Response (first pass - needs cleanup)

quux00 committed:
Clean up for first draft PR review

quux00 committed:
Update docs/changelog/112472.yaml

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:
```diff
@@ -151,8 +167,54 @@ public IndexResolution mergedMappings(String indexPattern, FieldCapabilitiesResp
+                    // MP FIXME this causes a 500 - is that what we want to return? _search returns 400 when skip_unavailable=false.
+                    throw new IllegalStateException(
+                        "Placeholder error for failing an ESQL search. Required remote cluster could not be contacted:" + clusterAlias
+                    );
```
Adding this here now to show how it will work, but the code to handle skip_unavailable (including throwing this error here) will be done in a follow on PR, so I will remove this after the initial draft review. But I wanted to include it here so you can see where it fits in with the execution info design.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java:
```diff
@@ -29,30 +32,71 @@
+    private final EsqlExecutionInfo esqlExecutionInfo;
+
+    // for use by top level ComputeListener
+    ComputeListener(
```
I think we should make a `private` ctor and make these named static methods.

quux00 replied:
Thanks Nik. Change made in [this commit](https://github.com/elastic/elasticsearch/pull/112472/commits/cd37244aab6fdb85d911da32db644c5fff70f296).

quux00 committed:
Restructured CrossClustersQueryIT so that num shards can be obtained; added new test that is failing when remote index missing

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java:
```diff
@@ -73,13 +117,75 @@ ActionListener<Void> acquireAvoid() {
+     * phase. Per-cluster took time is recorded in the {@link EsqlExecutionInfo}.
+     * @param clusterAlias remote cluster alias the data node compute is running on
+     */
+    ActionListener<ComputeResponse> acquireComputeForDataNodes(String clusterAlias, Configuration configuration) {
```
What is the difference between acquireComputeForDataNodes with and without arguments? Can we streamline them?

quux00 replied:
Two part answer:

First, I didn't mean to rename the original `acquireCompute`. I see now that IntelliJ grabbed both when I did a rename of the one that takes args. I renamed it back in [this commit](https://github.com/elastic/elasticsearch/pull/112472/commits/1aea20f6e7043ed3f99d112517ba0989efe6f28f).

Second, you might still ask - why do we need `acquireCompute`, `acquireComputeForDataNodes` and `acquireCCSCompute` now. The reason is that `acquireCompute` was used at various levels of the "compute" phases and sub-phases (I'm not sure the right ESQL terminology for this yet). I need some spot or hook where I know that higher level results are coming back - either the final response from a remote cluster (`acquireCCSCompute`) or each "final" result from each data node - so this is set at the top of the "compute on data nodes phase" (starts in `startComputeOnDataNodes`) in order to be able to calculate per-cluster took time. 

All the other "sub-phases" or sub-computes use the original `acquireCompute`, and those phases (AFAICT) hold no useful information for me in terms of gathering execution metadata.

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java:
```diff
@@ -73,13 +117,75 @@ ActionListener<Void> acquireAvoid() {
+            TimeValue tookOnDataNode = new TimeValue(tookTimeMillis);
+            esqlExecutionInfo.swapCluster(clusterAlias, (k, v) -> {
+                if (v.getTook() == null || v.getTook().millis() < tookOnDataNode.millis()) {
+                    return new EsqlExecutionInfo.Cluster.Builder(v).setStatus(EsqlExecutionInfo.Cluster.Status.SUCCESSFUL)
```
This doesn't seem right. It looks like we received a response from only one node, rather than from the entire cluster.

quux00 replied:
Yes, good catch. I forgot to remove the setting of SUCCESSFUL status here when doing code cleanup. That status gets properly set elsewhere. I removed that in [this commit](https://github.com/elastic/elasticsearch/pull/112472/commits/09fa22441a860e4e22b1983382aa2138afd53406), but left in the took time calculation code.

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java:
```diff
@@ -513,6 +532,7 @@ private void lookupDataNodes(
         QueryBuilder filter,
         Set<String> concreteIndices,
         OriginalIndices originalIndices,
+        EsqlExecutionInfo executionInfo,
```
Can we avoid mutating executionInfo and instead return a list of data nodes and shard information?

quux00 replied:
Good idea. I have made a go of that in [this commit](https://github.com/elastic/elasticsearch/pull/112472/commits/29b996bf96249c9a12561f6ad36846fd08511cec). Let me know what you think of that approach instead.

dnhatn reviewed (commented):
Thanks, Michael. The approach looks good. However, I prefer minimizing the passing and mutation of ExecutionInfo. Instead, we could return the existing data alongside the new information. For example, lookupDataNodes() could return a list of data nodes and shard infos together. This would make the APIs easier to maintain.

quux00 committed:
Handled case where a cluster is specified with no matching indices.

In that case, the field-caps response returns nothing about that cluster
so we can't provide any metadata about the cluster. The fix involved
passing in the RemoteClusterService into EsqlSesssion so it can do the
cluster resolution before the field-caps call and have some basic initial
requested issue about each cluster specified even if no search is done
against that cluster because the index expression specified matches no
indices.

quux00 committed:
PR feedback - lookupDataNodes returns can_match info to fill in ExecutionInfo in the Listener rather than passing into lookupDataNodes method

quux00 committed:
PR feeback: removed incorrect setting of cluster status to successful in acquireComputeForDataNodes

quux00 committed:
Renamed acquireComputeForDataNodes with no args back to acquireCompute (in ComputeListener)

quux00 committed:
PR feedback: created static methods to create ComputeService objects; added javadoc for the new variant

quux00 commented:
Closed in favor of: https://github.com/elastic/elasticsearch/pull/112595
